File: src/hand.ml
open Card

type hand = card Dynarray.t

let playable_cards hand first_card =
  if Dynarray.exists (same_color first_card) hand then
    Dynarray.filter (same_color first_card) hand
  else hand

let copy_without hand card =
  let new_hand = Dynarray.create () in
  Dynarray.iter (fun c -> if c <> card then Dynarray.add_last new_hand c) hand;
  new_hand

File: rust/the-crew-solver/src/hand.rs
use std::ops::Deref;

use crate::card::Card;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Hand {
    cards: Vec<Card>,
}

impl Deref for Hand {
    type Target = Vec<Card>;

    fn deref(&self) -> &Self::Target {
        &self.cards
    }
}

impl Hand {
    pub fn playable_cards(&self, first_card: &Card) -> Vec<Card> {
        let same_color = |c: &Card| first_card.same_color(c);
        if self.cards.iter().any(same_color) {
            self.cards.iter().copied().filter(same_color).collect()
        } else {
            self.cards.clone()
        }
    }
}
